### DISCUSSION TITLE
Converting np.ndarrays to jnp.arrays causes memory leak

### DISCUSSION THREAD

**Discussion by jzhao321** [jax-ml/jax#14876] [Q&A]
I've been playing around on Google Colab with GPU instances.
When I call jax.device_put() on a converted numpy array, there seems to be memory allocated that cannot be recovered even after running jax.clear_backends() and gc.collect()

not sure if this is intended behavior, I wasn't able to find anything about this in the docs.

Jax version is one prepackaged with Google Colab

```python
import gc
from psutil import Process
import numpy as np
import jax.numpy as jnp
import jax

os.environ['XLA_PYTHON_CLIENT_PREALLOCATE'] = "false"
os.environ['XLA_PYTHON_CLIENT_ALLOCATOR'] = "platform"

def mem_profile():
  return print_mem(Process().memory_info().rss)

def print_mem(mem):
  mb = mem / (1024 * 1024)
  return "{:.2f} MB".format(mb)

def move_np_to_jnp(state):
  flattened, structure = jax.tree_util.tree_flatten(state)

  unsharded = []
  for flattened_array in flattened:
    jnp_array = jnp.asarray(flattened_array)
    unsharded.append(jax.device_put(jnp_array, device=jax.devices()[0]))

  state = jax.tree_util.tree_unflatten(structure, unsharded)
  return state
  
for index in range(10):
  print("Before Loop: ", index, mem_profile())
  state_1 = move_np_to_jnp(state = {
    "test_key_1": np.ones((1024,512,256)),
    "test_key_2": np.ones((1024,512,256))
  })

  state_2 = move_np_to_jnp(state = {
    "test_key_3": np.ones((1024,1024,128)),
    "test_key_4": np.ones((1024,1024,128))
  })

  del state_1, state_2
  jax.clear_backends()
  gc.collect()
  print("After Loop: ", index, mem_profile())

print("After All: ", mem_profile())
```

**Comment by jakevdp** [comment]
I think this is working as intended. The device does not grow its memory pool until it is needed, but it will never shrink its memory pool once it's grown. After the initial allocations, the memory pool never expands again, because new allocations fit in the existing pool.

Does that make sense?

Side note, your `move_to_jnp` function could be written more efficiently like this:
```python
def move_np_to_jnp(state):
  return jax.device_put(state)
```
(`device_put` natively handles pytrees)

**Reply by hawkinsp** [reply]
Hmm, I'm not sure. Since we have opted into the platform allocator, we should immediately release memory.

**Reply by jzhao321** [reply]
After running your suggested snippet, The memory allocation stopped increasing after every loop iteration.
I'm happy that it works now, however I'm confused on what's happening behind the scenes.
```
def move_np_to_jnp(state):
  flattened, structure = jax.tree_util.tree_flatten(state)

  unsharded = []
  for flattened_array in flattened:
    jnp_array = jnp.asarray(flattened_array)
    unsharded.append(jax.device_put(jnp_array, device=jax.devices()[0]))

  state = jax.tree_util.tree_unflatten(structure, unsharded)
  return state

def move_np_to_jnp_new(state):
  return jax.tree_map(jax.device_put, state)
```
Theoretically, these two functions serve the same purpose, however the more verbose function has a memory leak even with forced garbage collection every loop.

**Reply by jzhao321** [reply]
Also, I was hoping to find a method to clear out Jax's memory allocations and cache programmatically without having to restart the entire process. I came across `jax.clear_backends()`, however this doesn't seem to do anything in this situation. The memory usage after running `move_np_to_jnp` remains high.

**Reply by jzhao321** [reply]
I can verify through google colab's variable inspector that everything has been deleted using `del` and the garbage collector has ran multiple times.
I can also verify that it's not a race condition, where jax is "waiting" some time to deallocate the memory. After waiting several minutes then executing `mem_profile()` results in the same thing.

I'm using the standard RAM shape in google colab, and this is what the code block is able to output before the instance crashes (note, this is the second time running the block, that's why the init memory begins at 6GB):
```
Before Loop:  0 6077.71 MB
After Loop:  0 6605.83 MB
Before Loop:  1 6605.83 MB
After Loop:  1 7133.86 MB
Before Loop:  2 7133.86 MB
After Loop:  2 7661.91 MB
Before Loop:  3 7661.91 MB
After Loop:  3 8188.60 MB
Before Loop:  4 8188.60 MB
```

**Reply by jakevdp** [reply]
One thing to keep in mind: when a JAX array is garbage collected on the Python side, it calls `arr.delete()` in order to clear the device memory, and this is an asynchronous call that may take a few cycles to complete. If your Python program carries on and queries memory use immediately, you may be probing the memory before it's cleared.

**Reply by jzhao321** [reply]
How come using jax.tree_map() circumvents this issue?
Also, what @hawkinsp said is true, Jax should be deallocating memory buffers if the `XLA_PYTHON_CLIENT_ALLOCATOR` is set to `platform`
If the memory usage is coming from Jax preallocation, then it only works on the GPU devices, not CPU. Not sure if there's a way to disable that on the CPU.

**Reply by jzhao321** [reply]
Seems to be related to this https://github.com/google/jax/issues/12340

**Reply by jzhao321** [reply]
Also, after I call device_put() on a numpy array, shouldn't the CPU memory be freed since the data is now backed by the GPU buffer?
When running this snippet:
```
print("Before", mem_profile())

state = {
  "test_key_1": np.ones((1024,512,256), dtype=np.float16),
  "test_key_2": np.ones((1024,512,256), dtype=np.float16)
}
state = jax.tree_map(jax.tree_util.Partial(jax.device_put, device=jax.devices()[0]), state)

gc.collect()
print("After", mem_profile())
```
I get this output:
```
Before 846.15 MB
After 1373.54 MB
```

**Reply by jakevdp** [reply]
Sorry, I guess I'm confused here... are you concerned about JAX's on-device memory allocation, or numpy's on-host memory allocation? Your last example here seems like it's probing the latter. You could replace the sixth line with `state = None`, removing JAX entirely from the snippet, and see the same results, no?

**Reply by jzhao321** [reply]
Apologies for the confusion, my english is not too great.

The specific use case I'm asking for is Loading sharded Flax checkpoints for inferencing large models such as OPT 6.7b.
Since loading the entire checkpoint would blow out the memory in Google Colab, I was hoping to load each file individually, moving it to the GPU, then deleting the file from the CPU host memory before moving onto the next file.
The Msgpacked flax checkpoints become numpy arrays when deserialized.

Regarding the `state = None` you suggested, If I replace `state = jax.tree_map(...)` with `state = None`
This is the output:
```
Before 148.57 MB
After 148.63 MB
```
This shows that the memory allocation increasing is a result of JAX doing something behind the scenes.

**Reply by jzhao321** [reply]
Note that this is not an issue if i declare the data directly using JAX numpy:
```
print("Before", mem_profile())

state = jax.device_put(jnp.ones((1024,1024,1024)), device=jax.devices()[0])

gc.collect()
print("After", mem_profile())
```
```
Before 1181.70 MB
After 1186.30 MB
```
This only seems to happen when converting some other data format to JAX, then putting on a device.

EDIT:
Actually, no when testing again I realized that the memory usage has also spiked up after declaring directly using `jnp.ones(...)`
It seems JAX holds a copy of the data on both CPU and GPU.

**Reply by jakevdp** [reply]
I see, thanks. Looking at refcounts more directly, it looks like you're right: something in `jax.device_put` is keeping a reference to the numpy array:
```python
import sys
import jax
import numpy as np

x = np.arange(1000, dtype='float32')
print(sys.getrefcount(x))
# 2

y = jax.device_put(x)
print(sys.getrefcount(x))
# 3

del y
print(sys.getrefcount(x))
# 3
```

**Reply by jakevdp** [reply]
This looks like a real bug; I've opened https://github.com/google/jax/issues/14882 to track it. Thanks for raising this!

**Reply by jzhao321** [reply]
Thank God, i wasn't going crazy.

**Reply by jzhao321** [reply]
@jakevdp I'm not familiar with the internals of JAX.
While this bug is being worked on, do you know of a workaround?

**Reply by jakevdp** [reply]
See the bug for a suggested solution to calling XLA's internal garbage collection.

I think your snippet is suffering from something else as well, because you're creating two on-device copies of each array (`jnp_array = jnp.asarray(flattened_array)` pushes one copy to the default device, then `unsharded.append(jax.device_put(jnp_array, device=jax.devices()[0]))` creates another copy)

If I replace `move_np_to_jnp` with a simple `jax.device_put` and manually call the xla garbage collection, I don't see the memory growth any more:
```python
import os
os.environ['XLA_PYTHON_CLIENT_PREALLOCATE'] = "false"
os.environ['XLA_PYTHON_CLIENT_ALLOCATOR'] = "platform"

import gc
from psutil import Process
import numpy as np
import jax.numpy as jnp
import jax
from jax._src.lib import xla_client

def mem_profile():
  return print_mem(Process().memory_info().rss)

def print_mem(mem):
  mb = mem / (1024 * 1024)
  return "{:.2f} MB".format(mb)
  
for index in range(5):
  print("Before Loop: ", index, mem_profile())
  state_1 = jax.device_put({
    "test_key_1": np.ones((1024,512,256)),
    "test_key_2": np.ones((1024,512,256))
  })

  state_2 = jax.device_put({
    "test_key_3": np.ones((1024,1024,128)),
    "test_key_4": np.ones((1024,1024,128))
  })

  del state_1, state_2
  jax.clear_backends()
  gc.collect()
  xla_client._xla.collect_garbage()
  print("After Loop: ", index, mem_profile())

print("After All: ", mem_profile())
```
Output:
```
Before Loop:  0 150.04 MB
After Loop:  0 1291.35 MB
Before Loop:  1 1291.35 MB
After Loop:  1 779.68 MB
Before Loop:  2 779.67 MB
After Loop:  2 267.71 MB
Before Loop:  3 267.71 MB
After Loop:  3 267.71 MB
Before Loop:  4 267.71 MB
After Loop:  4 267.72 MB
After All:  267.72 MB
```
Note that the XLA garbage collection is asynchronous, which explains why the memory jumps up and down.

**Reply by jzhao321** [reply]
@jakevdp I have a follow up question, not sure if topical.
When running just this:
```
state_1 = jax.device_put({
    "test_key_1": np.ones((1024,512,256)),
    "test_key_2": np.ones((1024,512,256))
  })
```
I see in colab the CPU and GPU memory both go up, suggesting that JAX is storing the data on both. Is this intended behavior?
If so, I don't see how I would be able to load OPT 6.7B (13.31 gb fp16) in colab's limited 12.7 gb CPU buffer.

**Reply by jakevdp** [reply]
Do you have X64 mode enabled? If not, then pushing 64-bit arrays to JAX will result in a 32-bit copy on host-side, along with a 32-bit copy on device. To avoid this you can either enable X64 mode, or you can use 32-bit source arrays.

**Reply by jzhao321** [reply]
@jakevdp It seems to be allocating regardless of X64 mode and dtype of the numpy array.
It also seems to be happening even with data defined by `jnp.ones(...)`

EDIT:
I can't seem to reproduce CPU memory allocation with `jnp.ones(...)` after Deleting / Starting a new runtime in Colab.

EDIT 2:
There's definitely some weird behavior going on. JAX memory allocation seems to be inconsistant with `jnp.ones(...)`. Sometimes it takes up CPU ram, sometimes it doesn't. It changes between runs. Haven't figured out what causes it yet.
Running `jax.device_put(...)` with numpy arrays always allocate CPU memory, even if declared inline.

EDIT 3:
When declaring using `jnp.ones(...)`, it "consumes" cpu memory because I had the Colab Variable inspector open. The variable inspector forces jax to fetch the values to cpu ram, thus explaining the memory consumption.

However, when declaring a numpy array then calling device.put() on the value, it still consumes CPU memory even after removing the reference.

```
import os
os.environ['XLA_PYTHON_CLIENT_PREALLOCATE'] = "false"
os.environ['XLA_PYTHON_CLIENT_ALLOCATOR'] = "platform"
os.environ['JAX_ENABLE_X64'] = "True"
import gc
from psutil import Process
import jax.numpy as jnp
import jax
from jax._src.lib import xla_client
import numpy as np
import time

def mem_profile():
  return print_mem(Process().memory_info().rss)

def print_mem(mem):
  mb = mem / (1024 * 1024)
  return "{:.2f} MB".format(mb)

def clear_mem():
  jax.clear_backends()
  xla_client._xla.collect_garbage()
  gc.collect()

# Test 1
print("Baseline", mem_profile())

state_1 = jax.device_put(jnp.ones((1024,1024,1024), dtype=jnp.float16), device=jax.devices()[0])
time.sleep(5) # Ensure Async ops are executed
print("After Init JNP", mem_profile())

del state_1
clear_mem()
time.sleep(5) # Ensure Async ops are executed
print("Cleared JNP", mem_profile())

# Test 2
print("Baseline 2", mem_profile())

state_2_numpy = {
  "test_key_1": np.ones((1024,1024,512), dtype=np.float16),
  "test_key_2": np.ones((1024,1024,512), dtype=np.float16)
}
state_2 = jax.device_put(state_2_numpy, device=jax.devices()[0])
time.sleep(5) # Ensure Async ops are executed
print("After Init Numpy Dict", mem_profile())

del state_2_numpy
clear_mem()
time.sleep(5) # Ensure Async ops are executed
print("Cleared Numpy Dict On CPU", mem_profile())

del state_2
clear_mem()
time.sleep(5) # Ensure Async ops are executed
print("Cleared Numpy Dict On GPU", mem_profile())

# Test 3
print("Baseline 3", mem_profile())

state_3 = jax.device_put({
  "test_key_1": np.ones((1024,1024,512), dtype=np.float16),
  "test_key_2": np.ones((1024,1024,512), dtype=np.float16)
}, device=jax.devices()[0])
time.sleep(5) # Ensure Async ops are executed
print("After Init Numpy Dict Inline", mem_profile())

del state_3
clear_mem()
time.sleep(5) # Ensure Async ops are executed
print("Cleared Numpy Dict Inline", mem_profile())

# Test 4
print("Baseline 4", mem_profile())

state_4_numpy = np.ones((1024,1024,1024), dtype=np.float16)
state_4 = jax.device_put(state_4_numpy, device=jax.devices()[0])
time.sleep(5) # Ensure Async ops are executed
print("After Init Numpy Array", mem_profile())

del state_4_numpy
clear_mem()
time.sleep(5) # Ensure Async ops are executed
print("Cleared Numpy Array On CPU", mem_profile())

del state_4
clear_mem()
time.sleep(5) # Ensure Async ops are executed
print("Cleared Numpy Array On GPU", mem_profile())

# Test 5
print("Baseline 5", mem_profile())

state_5 = jax.device_put(np.ones((1024,1024,1024), dtype=np.float16), device=jax.devices()[0])
time.sleep(5) # Ensure Async ops are executed
print("After Init Numpy Array Inline", mem_profile())

del state_5
clear_mem()
time.sleep(5) # Ensure Async ops are executed
print("Cleared Numpy Array Inline", mem_profile())
```
Output:
```
Baseline 1126.88 MB
After Init JNP 1130.17 MB
Cleared JNP 1130.19 MB
Baseline 2 1130.19 MB
After Init Numpy Dict 5290.35 MB
Cleared Numpy Dict On CPU 3242.23 MB
Cleared Numpy Dict On GPU 1130.21 MB
Baseline 3 1130.21 MB
After Init Numpy Dict Inline 5290.37 MB
Cleared Numpy Dict Inline 1130.21 MB
Baseline 4 1130.21 MB
After Init Numpy Array 5290.14 MB
Cleared Numpy Array On CPU 3242.23 MB
Cleared Numpy Array On GPU 1130.21 MB
Baseline 5 1130.21 MB
After Init Numpy Array Inline 5290.15 MB
Cleared Numpy Array Inline 1130.22 MB
```
From this, something is definetly taking up memory on the CPU as a result of JAX ops behind the scenes.

**Reply by jzhao321** [reply]
Output from nvidia-smi:
```
Fri Mar 10 00:36:11 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.85.12    Driver Version: 525.85.12    CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   78C    P0    34W /  70W |    361MiB / 15360MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
```

Output from `jax.devices()`
```
[StreamExecutorGpuDevice(id=0, process_index=0, slice_index=0)]
```

**Reply by jzhao321** [reply]
@jakevdp Problem can be illustrated more though this snippet:
```
import os
os.environ['XLA_PYTHON_CLIENT_PREALLOCATE'] = "false"
os.environ['XLA_PYTHON_CLIENT_ALLOCATOR'] = "platform"
os.environ['JAX_ENABLE_X64'] = "True"
import gc
from psutil import Process
import jax.numpy as jnp
import jax
from jax._src.lib import xla_client
import numpy as np
import time

print("Jax Version:", jax.__version__)

def mem_profile():
  return print_mem(Process().memory_info().rss)

def print_mem(mem):
  mb = mem / (1024 * 1024)
  return "{:.2f} MB".format(mb)

def clear_mem():
  jax.clear_backends()
  xla_client._xla.collect_garbage()
  gc.collect()

for index in range(10):
  print("Baseline", index, mem_profile())
  state_4_numpy = np.ones((512,512,512), dtype=np.float16)
  state_4_jax_numpy = jnp.asarray(state_4_numpy, dtype=jnp.float16)
  state_4 = jax.device_put(state_4_jax_numpy, device=jax.devices()[0])
  print("After Init", mem_profile())

  del state_4_numpy, state_4_jax_numpy, state_4
  clear_mem()
  time.sleep(10) # Ensure Async ops are executed
  print("After Delete", mem_profile())
```
Output:
```
Jax Version: 0.4.6
Baseline 0 150.53 MB
After Init 772.03 MB
After Delete 772.26 MB
Baseline 1 772.26 MB
After Init 1036.50 MB
After Delete 1036.45 MB
Baseline 2 1036.45 MB
After Init 1300.70 MB
After Delete 1300.61 MB
Baseline 3 1300.61 MB
After Init 1564.86 MB
After Delete 1564.76 MB
Baseline 4 1564.76 MB
After Init 1831.33 MB
After Delete 1831.24 MB
Baseline 5 1831.24 MB
After Init 2095.52 MB
After Delete 2095.43 MB
Baseline 6 2095.43 MB
After Init 2359.67 MB
After Delete 2359.62 MB
Baseline 7 2359.62 MB
After Init 2623.86 MB
After Delete 2623.77 MB
Baseline 8 2623.77 MB
After Init 2888.05 MB
After Delete 2887.96 MB
Baseline 9 2887.96 MB
After Init 3154.54 MB
After Delete 3154.47 MB
```
